src/SqlFormatter.php

```diff
@@ -198,8 +187,8 @@ public function format(string $string, string $indentString = '  ') : string
                 }
 
                 // Take out the preceding space unless there was whitespace there in the original query
-                if (isset($originalTokens[$indexedToken['originalIndex']-1]) &&
-                    ! $originalTokens[$indexedToken['originalIndex']-1]->isOfType(Token::TOKEN_TYPE_WHITESPACE)) {
+                $prevToken = $cursor->subCursor()->previous();
+                if ($prevToken && ! $prevToken->isOfType(Token::TOKEN_TYPE_WHITESPACE)) {
```

Comment:
I'm not requiring you do it, but how would you feel about implementing the null object pattern to avoid all these checks?

Reply:
hmm. not sure about that. to do so i will have to introduce a `TokenInterface` and also to use an `instanceof` check in the while, or add one more method in the `Cursor` class... do you have a suggestion on how to implement it?

Reply:
I would rather add a new type, and check it in most methods if needed (but `isOfType` would not need to change, for instance). `value()` would require throwing, so would `withValue`

Reply:
Ah drop it, this is probably just another bad design idea on my part :sweat: